File: 10_Graphs/Shortest Path/Dijkstra's/Dijkstra.cpp
#include <iostream>
#include <vector>
#include <set>
#include <climits>

using namespace std;

class Graph {
    int V;
    vector<vector<pair<int, int>>> adj;

public:
    Graph(int v) : V(v) {
        adj.resize(V);
    }

    void addEdge(int u, int v, int w) {
        adj[u].emplace_back(v, w);
        adj[v].emplace_back(u, w);  // For an undirected graph
    }

    void dijkstra(int src) {
        vector<int> dist(V, INT_MAX);
        set<pair<int, int>> setds;

        setds.insert({0, src});
        dist[src] = 0;

        while (!setds.empty()) {
            int u = setds.begin()->second;
            setds.erase(setds.begin());

            for (auto &[v, weight] : adj[u]) {
                if (dist[u] + weight < dist[v]) {
                    if (dist[v] != INT_MAX)
                        setds.erase(setds.find({dist[v], v}));

                    dist[v] = dist[u] + weight;
                    setds.insert({dist[v], v});
                }
            }
        }

        cout << "Vertex Distance from Source (" << src << ")" << endl;
        for (int i = 0; i < V; ++i)
            cout << i << "\t\t" << dist[i] << endl;
    }
};

int main() {
    Graph g(5);
    g.addEdge(0, 1, 9);
    g.addEdge(0, 2, 6);
    g.addEdge(0, 3, 5);
    g.addEdge(0, 4, 3);
    g.addEdge(2, 1, 2);
    g.addEdge(2, 3, 4);

    g.dijkstra(0);

    return 0;
}

File: 10_Graphs/Shortest Path/Dijkstra's/Dijkstra.py
import heapq
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v, w):
        self.graph[u].append((v, w))
        self.graph[v].append((u, w))  # For an undirected graph

    def dijkstra(self, src):
        pq = [(0, src)]
        dist = {node: float('inf') for node in self.graph}
        dist[src] = 0

        while pq:
            current_distance, u = heapq.heappop(pq)

            if current_distance > dist[u]:
                continue

            for v, weight in self.graph[u]:
                distance = current_distance + weight

                if distance < dist[v]:
                    dist[v] = distance
                    heapq.heappush(pq, (distance, v))

        print(f"Vertex Distance from Source ({src})")
        for node in self.graph:
            print(f"{node}\t\t{dist[node]}")

# Sample usage
g = Graph()
g.add_edge(0, 1, 9)
g.add_edge(0, 2, 6)
g.add_edge(0, 3, 5)
g.add_edge(0, 4, 3)
g.add_edge(2, 1, 2)
g.add_edge(2, 3, 4)

g.dijkstra(0)
